compiler/rustc_mir/src/transform/basic_block_copy_prop.rs

```diff
@@ -0,0 +1,366 @@
+    /// deallocated or modified.
+    invalidation_map: IndexVec<Local, Invalidation>,
+
+    /// Data generation in this map.
+    ///
+    /// This is bumped when entering a new block. When looking data up, we ensure it was stored in
+    /// the same generation. This allows clearing the map by simply incrementing the generation
+    /// instead of having to clear the data (which can perform poorly).
+    generation: u32,
+}
+
+impl<'tcx> LocalValues<'tcx> {
+    fn new(body: &Body<'_>) -> Self {
+        Self {
+            map: IndexVec::from_elem_n(LocalValue::Unknown, body.local_decls.len()),
+            invalidation_map: IndexVec::from_elem_n(
+                Invalidation::default(),
+                body.local_decls.len(),
+            ),
+            generation: 0,
+        }
+    }
+
+    fn get(&self, local: Local) -> Option<Place<'tcx>> {
+        match self.map[local] {
+            LocalValue::Place { place, generation } if generation == self.generation => Some(place),
+            _ => None,
+        }
+    }
+
+    /// Records an assignment of `place` to `local`.
+    fn insert(&mut self, local: Local, place: Place<'tcx>) {
+        self.map[local] = LocalValue::Place { place, generation: self.generation };
+
+        let inval = &mut self.invalidation_map[place.local];
+        if inval.generation != self.generation {
+            inval.locals.clear();
+            inval.generation = self.generation;
+        }
+
+        inval.locals.push(local);
+    }
+
+    /// Resets `local`'s state to `Unknown` and invalidates all locals that have been assigned to
+    /// `local` in the past.
+    fn invalidate(&mut self, local: Local) {
+        self.map[local] = LocalValue::Unknown;
+
+        let inval = &mut self.invalidation_map[local];
+        if inval.generation == self.generation {
+            for local in &inval.locals {
+                self.map[*local] = LocalValue::Unknown;
+            }
+
+            inval.locals.clear();
+        }
+    }
+
+    /// Marks the data in the map as dirty, effectively clearing it.
+    fn clear(&mut self) {
+        self.generation += 1;
+    }
+}
+
+struct CopyPropVisitor<'tcx> {
+    tcx: TyCtxt<'tcx>,
+
+    /// Locals that have their address taken.
+    referenced_locals: BitSet<Local>,
+
+    /// Tracks the (symbolic) values of local variables in the visited block.
+    local_values: LocalValues<'tcx>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for CopyPropVisitor<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_statement(&mut self, statement: &mut Statement<'tcx>, location: Location) {
+        // We are only looking for *copies* of a place, not moves, since moving out of the place
+        // *again* later on would be a use-after-move. For example:
+        //   _1 = ...;
+        //   _2 = move _1;
+        //   use(move _2);   <- can *not* replace with `use(move _1)` or `use(_1)`
```

Comment:
Is this expected to get changed in the future if `_2` is only ever used right here?